Tactics/StepThms.lean

```diff
@@ -33,11 +41,190 @@ initialize registerTraceClass `gen_step.print_names
+given that `state.program = program` -/
+private def fetchLemma (state program h_program : Expr)
+    (addr : BitVec 64) (rawInst : BitVec 32) : Expr :=
+  let someRawInst := toExpr (some rawInst)
+  mkAppN (mkConst ``fetch_inst_eq_of_prgram_eq_of_map_find) #[
+    state,
+    program,
+    toExpr addr,
+    someRawInst,
+    h_program,
+    mkApp2 (.const ``Eq.refl [1])
+      (mkApp (.const ``Option [0]) <|
+        mkApp (.const ``BitVec []) (toExpr 32))
+      someRawInst
+  ]
+
+-- /-! ## `reduceDecodeInst` -/
+
+/-- `canonicalizeBitVec e` recursively walks over expression `e` to convert any
+occerences of:
+  `BitVec.ofFin w (Fin.mk x _)`
+to the canonical form:
+  `BitVec.ofNat w x` (i.e., `x#w`)
+ -/
+-- TODO: should this canonicalize to `BitVec.ofNatLt` instead,
+--       as the current transformation loses information?
+partial def canonicalizeBitVec (e : Expr) : MetaM Expr := do
+  match_expr e with
+    | BitVec.ofFin w i =>
+        let_expr Fin.mk _ x _h := i | fallback
+        let w ←
+          if w.hasFVar || w.hasMVar then
+            pure w
+          else
+            withTransparency .all <| reduce w
+            -- ^^ NOTE: potentially expensive reduction
+        return mkApp2 (mkConst ``BitVec.ofNat) w x
+    | _ => fallback
+  where
+    fallback : MetaM Expr := do
+      let fn   := e.getAppFn
+      let args  ← e.getAppArgs.mapM canonicalizeBitVec
+      return mkAppN fn args
+
+/-- Given an expr `rawInst` of type `BitVec 32`,
+return an expr of type `Option ArmInst` representing what `rawInst` decodes to.
+The resulting expr is guaranteed to be def-eq to `fetch_inst $rawInst` -/
+def reduceDecodeInstExpr (rawInst : Expr) : MetaM Expr := do
+  let expr := mkApp (mkConst ``decode_raw_inst) rawInst
+  let expr ← withTransparency .all <| reduce expr
+  -- ^^ NOTE: possibly expensive reduction
+  canonicalizeBitVec expr
+
+/-! ## SymM Monad -/
+
+abbrev SymM.CacheKey := BitVec 32
+abbrev SymM.CacheM := MonadCacheT CacheKey Expr MetaM
+abbrev SymM := ProgramInfoT <| MonadCacheT SymM.CacheKey Expr MetaM
+
+@[inherit_doc ProgramInfoT.run]
+abbrev SymM.run (name : Name) (k : SymM α) (persist : Bool := true) : MetaM α :=
+  MonadCacheT.run <| ProgramInfoT.run name k persist
+
+open SymM in
+/-- Given a (reflected) raw instruction,
+return an expr of type `Option ArmInst` representing what `rawInst` decodes to.
+The resulting expr is guaranteed to be def-eq to `fetch_inst $rawInst`.
+
+Results are cached so that the same instruction is not reduced multiple times -/
+def reduceDecodeInst (rawInst : BitVec 32) : CacheM Expr :=
+  checkCache (rawInst) fun _ =>
+    reduceDecodeInstExpr (toExpr rawInst)
+
+open ProgramInfoT InstInfoT
+
+/-! ## reduceStepiToExecInst -/
+
+/-- Given a program and an address, and optionally the corresponding
+raw and decoded instructions, construct and return first the expression:
+```
+∀ {s} (h_program : s.program = <progam>) (h_pc : read_pc s = <addr>)
+  (h_err : read_err s = .None),
```

Comment:
```suggestion
∀ {s} (h_program : s.program = <progam>) (h_pc : r .PC s = <addr>)
  (h_err : r .ERR s = .None),
```

Reply:
Something to keep in mind: some LDST instructions expect the SP to be aligned (e.g., https://github.com/leanprover/LNSym/blob/6d1ef32d9d253b8b7667edfaa08f38ea82da2dc1/Arm/Insts/LDST/Reg_unscaled_imm.lean#L28). We may need that assumption for such instructions.

Reply:
Yes, this is the same problem as you mentioned for branch: the exact conditions needed for a semantics differ. For now, such conditions will remain present as `if`s in the simplified semantics, which are then reduced away by `sym_n`.

In the future, I'd like to make this smarter, by introducing a more elaborate datastructure that tracks exactly which assumptions are needed, and fully reduce/reflect for that set of assumptions. I figured this PR is already big enough as it is, so I didn't want to go too deep into that train of thought yet